### DISCUSSION TITLE
What's the best way to implement tags?

### DISCUSSION THREAD

**Discussion by boogerlad** [agentm/project-m36#343] [Q&A] [answered]
I see the most common approach for SQL databases is this: https://stackoverflow.com/a/20871/1201238
using three relations. One for the item, one for the tag, and one to link the previous two. (pretty much a many to many relationship)

However, with PostgreSQL specifically, it seems that an inline text array with a GIN index is the best: http://www.databasesoup.com/2015/01/tag-all-things-part-3.html

Seeing that Project:M36 is totally different, thought I should ask to see if there is a method more optimal before I implement it naively

**Comment by agentm** [comment, marked as answer]
As you've discovered, there are bunch of different options here:

* attribute of list of strings: this results in lots of duplicate tag name data, but it's easy-to-understand. Renaming tags or getting a complete list of tags is not easy, though. Project:M36 does not yet support the equivalent of GIN indexing.
* an attribute of list/set of a tag type: if your list of tags is static, then you could represent it through an ADT (or ENUM in other databases)- `data MyTag = TrucksTag | CarsTag | TrainsTag` 
* an attribute of list/set of a tag type w/ custom tags: like the option just above, but with the addition of a custom tag value `data MyTag = TrucksTag | CarsTag | TrainsTag | CustomTag String`- provides for easy matching on the most common tags but with the extensibility of additional tags. The downside is that not all tags are treated or can be queried the same way.
* a many-to-many relationship is probably the most flexible: the list of tags can be found in a single relation, renaming tags is trivial, matching tags to entries is just a join away which is natural for the database. This option is likely the most relation-algebra-friendly.

**Reply by boogerlad** [reply]
Thanks for this! I didn't know there was a list / set datatype in Project:M36. Is there any documentation? I'll probably stick with the many-to-many relationship for now.
